Cannot de-serialize serializable JDK proxy

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/fury/issues) and found no similar issues.


### Version

Fury: 0.71
JDK: 21

### Component(s)

Java

### Minimal reproduce step

```java
import org.apache.fury.Fury;
import org.apache.fury.config.Language;
import org.junit.jupiter.api.Test;

import java.io.ObjectStreamException;
import java.io.Serializable;
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

import static org.junit.jupiter.api.Assertions.assertEquals;

class ProxySerializerTest {

	@Test
	void serializeProxy() {
		final Fury fury = Fury.builder()
				.withLanguage(Language.JAVA)
				.requireClassRegistration(false)
				.build();

		final Object o = ProxyFactory.createProxy(TestInterface.class);
		final byte[] s = fury.serialize(o);
		assertEquals(fury.deserialize(s), o);
	}

	interface TestInterface {
		void test();
	}

	static class ProxyFactory {

		static <T> T createProxy(final Class<T> type) {
			return new JdkProxyFactory().createProxy(type);
		}

		static final class JdkProxyFactory {

			@SuppressWarnings("unchecked")
			<T> T createProxy(final Class<T> type) {
				final JdkHandler handler = new JdkHandler(type);
				try {
					final ClassLoader cl = Thread.currentThread().getContextClassLoader();
					return (T) Proxy.newProxyInstance(cl, new Class[]{type, Serializable.class}, handler);
				} catch (IllegalArgumentException e) {
					throw new RuntimeException("Could not create proxy for type [" + type.getName() + "]", e);
				}
			}

			static class JdkHandler implements InvocationHandler, Serializable {

				private final String typeName;

				private JdkHandler(Class<?> type) {
					typeName = type.getName();
				}

				@Override
				public Object invoke(Object proxy, Method method, Object[] args) {
					return null;
				}

				public Object writeReplace() throws ObjectStreamException {
					return new ProxyReplacement(typeName);
				}
			}

			static final class ProxyReplacement implements Serializable {

				private final String type;

				ProxyReplacement(final String type) {
					this.type = type;
				}

				private Object readResolve() throws ObjectStreamException {
					final Class<?> clazz;
					try {
						clazz = Class.forName(type, false, Thread.currentThread().getContextClassLoader());
					} catch (ClassNotFoundException ex) {
						throw new RuntimeException(ex);
					}
					return ProxyFactory.createProxy(clazz);
				}
			}
		}
	}
}
```


### What did you expect to see?

Serializable JDK proxies can be de-serialized.

### What did you see instead?

The following exception is thrown:

> java.lang.ClassCastException: class xyz.$Proxy8 cannot be cast to class java.lang.reflect.InvocationHandler

The problem seems to be the `JdkProxySerializer` and the `ReplaceResolveSerializer` get in each other's way because the `InvocationHandler` implements `writeReplace`.


### Anything Else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

Hi @theigl , I used JDK serialization to serialize your object, it throw ClassCastException too:
```java
java.lang.RuntimeException: java.lang.ClassCastException: cannot assign instance of org.apache.fury.serializer.$Proxy6 to field java.lang.reflect.Proxy.h of type java.lang.reflect.InvocationHandler in instance of org.apache.fury.serializer.$Proxy6

	at org.apache.fury.FuryTestBase.jdkDeserialize(FuryTestBase.java:339)
	at org.apache.fury.serializer.JdkProxySerializerTest.serializeProxy(JdkProxySerializerTest.java:142)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:136)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:658)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:219)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:923)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:192)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
	at org.testng.TestRunner.privateRun(TestRunner.java:808)
	at org.testng.TestRunner.run(TestRunner.java:603)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:429)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:423)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:383)
	at org.testng.SuiteRunner.run(SuiteRunner.java:326)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:95)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1249)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1169)
	at org.testng.TestNG.runSuites(TestNG.java:1092)
	at org.testng.TestNG.run(TestNG.java:1060)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:105)
Caused by: java.lang.ClassCastException: cannot assign instance of org.apache.fury.serializer.$Proxy6 to field java.lang.reflect.Proxy.h of type java.lang.reflect.InvocationHandler in instance of org.apache.fury.serializer.$Proxy6
	at java.base/java.io.ObjectStreamClass$FieldReflector.setObjFieldValues(ObjectStreamClass.java:2102)
	at java.base/java.io.ObjectStreamClass$FieldReflector.checkObjectFieldValueTypes(ObjectStreamClass.java:2066)
	at java.base/java.io.ObjectStreamClass.checkObjFieldValueTypes(ObjectStreamClass.java:1355)
	at java.base/java.io.ObjectInputStream$FieldValues.defaultCheckFieldValues(ObjectInputStream.java:2682)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2483)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2269)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1747)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:525)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:483)
	at org.apache.fury.FuryTestBase.jdkDeserialize(FuryTestBase.java:337)
	... 26 more
```
Here is my code for JDK serialization invocation:
```java
 public static byte[] jdkSerialize(Object o) {
    ByteArrayOutputStream bos = new ByteArrayOutputStream();
    try (ObjectOutputStream objectOutputStream = new ObjectOutputStream(bos)) {
      objectOutputStream.writeObject(o);
      objectOutputStream.flush();
      return bos.toByteArray();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static Object jdkDeserialize(byte[] data) {
    try (ByteArrayInputStream bis = new ByteArrayInputStream(data);
        ObjectInputStream objectInputStream =
            new ClassLoaderObjectInputStream(Thread.currentThread().getContextClassLoader(), bis)) {
      return objectInputStream.readObject();
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }
```

Seems `ProxyFactory.JdkProxyFactory.ProxyReplacement#readResolve`  should return `JdkHandler` instead of `Proxy` instance.

--- a/java/fury-core/src/main/java/org/apache/fury/resolver/ClassInfo.java
+++ b/java/fury-core/src/main/java/org/apache/fury/resolver/ClassInfo.java
@@ -114,7 +114,7 @@ public class ClassInfo {
     this.classId = classId;
     if (cls != null) {
       boolean isLambda = Functions.isLambda(cls);
-      boolean isProxy = ReflectionUtils.isJdkProxy(cls);
+      boolean isProxy = classId != ClassResolver.REPLACE_STUB_ID && ReflectionUtils.isJdkProxy(cls);
       this.isDynamicGeneratedClass = isLambda || isProxy;
       if (isLambda) {
         this.classId = ClassResolver.LAMBDA_STUB_ID;

--- a/java/fury-core/src/main/java/org/apache/fury/resolver/ClassResolver.java
+++ b/java/fury-core/src/main/java/org/apache/fury/resolver/ClassResolver.java
@@ -853,7 +853,11 @@ public class ClassResolver {
       } else if (Functions.isLambda(cls)) {
         return LambdaSerializer.class;
       } else if (ReflectionUtils.isJdkProxy(cls)) {
-        return JdkProxySerializer.class;
+        if (JavaSerializer.getWriteReplaceMethod(cls) != null) {
+          return ReplaceResolveSerializer.class;
+        } else {
+          return JdkProxySerializer.class;
+        }
       } else if (Calendar.class.isAssignableFrom(cls)) {
         return TimeSerializers.CalendarSerializer.class;
       } else if (ZoneId.class.isAssignableFrom(cls)) {

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/JdkProxySerializerTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/JdkProxySerializerTest.java
@@ -21,7 +21,9 @@ package org.apache.fury.serializer;
 
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertNotSame;
+import static org.testng.Assert.assertTrue;
 
+import java.io.ObjectStreamException;
 import java.io.Serializable;
 import java.lang.reflect.InvocationHandler;
 import java.lang.reflect.Method;
@@ -30,6 +32,7 @@ import java.util.function.Function;
 import org.apache.fury.Fury;
 import org.apache.fury.FuryTestBase;
 import org.apache.fury.config.Language;
+import org.apache.fury.reflect.ReflectionUtils;
 import org.testng.annotations.Test;
 
 @SuppressWarnings({"unchecked", "rawtypes"})
@@ -128,5 +131,92 @@ public class JdkProxySerializerTest extends FuryTestBase {
     RefTestInvocationHandler copyHandler =
         (RefTestInvocationHandler) Proxy.getInvocationHandler(copy);
     assertEquals(copyHandler.getProxy(), copy);
+  }
+
+  @Test
+  public void testSerializeProxyWriteReplace() {
+    final Fury fury =
+        Fury.builder().withLanguage(Language.JAVA).requireClassRegistration(false).build();
+
+    final Object o = ProxyFactory.createProxy(TestInterface.class);
+    final byte[] s = fury.serialize(o);
+    assertTrue(ReflectionUtils.isJdkProxy(fury.deserialize(s).getClass()));
+  }
+
+  interface TestInterface {
+    void test();
+  }
+
+  static class ProxyFactory {
+
+    static <T> T createProxy(final Class<T> type) {
+      return new JdkProxyFactory().createProxy(type);
+    }
+
+    public interface IWriteReplace {
+      Object writeReplace() throws ObjectStreamException;
+    }
+
+    static final class JdkProxyFactory {
+
+      @SuppressWarnings("unchecked")
+      <T> T createProxy(final Class<T> type) {
+        final JdkHandler handler = new JdkHandler(type);
+        try {
+          final ClassLoader cl = Thread.currentThread().getContextClassLoader();
+          return (T)
+              Proxy.newProxyInstance(
+                  cl, new Class[] {type, IWriteReplace.class, Serializable.class}, handler);
+        } catch (IllegalArgumentException e) {
+          throw new RuntimeException("Could not create proxy for type [" + type.getName() + "]", e);
+        }
+      }
+
+      static class JdkHandler implements InvocationHandler, IWriteReplace, Serializable {
+
+        private final String typeName;
+
+        private JdkHandler(Class<?> type) {
+          typeName = type.getName();
+        }
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+          if (isWriteReplaceMethod(method)) {
+            return writeReplace();
+          }
+          return null;
+        }
+
+        public Object writeReplace() throws ObjectStreamException {
+          return new ProxyReplacement(typeName);
+        }
+
+        static boolean isWriteReplaceMethod(final Method method) {
+          return (method.getReturnType() == Object.class)
+              && (method.getParameterTypes().length == 0)
+              && method.getName().equals("writeReplace");
+        }
+      }
+
+      public static final class ProxyReplacement implements Serializable {
+
+        private final String type;
+
+        public ProxyReplacement(final String type) {
+          this.type = type;
+        }
+
+        private Object readResolve() throws ObjectStreamException {
+          try {
+            final Class<?> clazz =
+                Class.forName(type, false, Thread.currentThread().getContextClassLoader());
+            return ProxyFactory.createProxy(clazz);
+          } catch (ClassNotFoundException ex) {
+            throw new RuntimeException(ex);
+          }
+        }
+      }
+    }
   }
 }

